File: 0367-valid-perfect-square/0367-valid-perfect-square.py
class Solution(object):
    def isPerfectSquare(self, num):
        if num < 2:
            return True
        left, right = 2, num // 2
        while left <= right:
            mid = (left + right) // 2
            sq = mid * mid
            if sq == num:
                return True
            elif sq < num:
                left = mid + 1
            else:
                right = mid - 1
        return False

File: 0367-valid-perfect-square/0367-valid-perfect-square.java
class Solution {
    public boolean isPerfectSquare(int num) {
        if (num < 2) return true;
        int left = 2;
        int right = num / 2;
        
        while (left <= right) {
            int mid = left + (right - left) / 2;
            long sq = (long) mid * mid; // prevent overflow
            if (sq == num) return true;
            else if (sq < num) left = mid + 1;
            else right = mid - 1;
        }
        
        return false;
    }
}
